### ISSUE TITLE
Incorrect warning about unread variable

### ISSUE THREAD

**Issue by shi-yan** [rust-lang/rust#77531]
I tried this code:

```rust
let new_state = EarleyState{rule:rule, start_position:loc, key:RuleKey{non_terminal_type:non_terminal::NonTerminalType::Init, index:scan_index}};
...
scan_index += 1;
```

I expected to see this happen: 

> The above code is in a loop. scan_index is a u32 type integer, I use it to create an object: RuleKey{..., index:scan_index} and then increase its value by one. Its value will be used next time when I create another RuleKey 

Instead, this happened: 

```The compiler generated this warning message, which is clearly wrong:
warning: value assigned to `scan_index` is never read
   --> src/parser/mod.rs:133:33
    |
133 | ...                   scan_index +=...
```

### Meta

`rustc --version --verbose`:
```
rustc 1.45.2 (d3fb005a3 2020-07-31)
binary: rustc
commit-hash: d3fb005a39e62501b8b0b356166e515ae24e2e54
commit-date: 2020-07-31
host: x86_64-apple-darwin
release: 1.45.2
LLVM version: 10.0```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>


warning: value assigned to `scan_index` is never read
   --> src/parser/mod.rs:133:33
    |
133 | ...                   scan_index +=...
    |                       ^^^^^^^^^^
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?


</p>
</details>

**Comment by jonas-schievink** [comment]
Please provide a more complete code example that incorrectly triggers the lint

**Comment by Dylan-DPC** [comment]
Closing this as this issue is vague and no further details were provided.
